sophia-vote-contract.md

```diff
@@ -0,0 +1,142 @@
+`````
+private function size(l : list('a)) : int = size'(l, 0)
+`````
+
+This is where things get a bit more complicated, so I will try to explain what is happening here.
+
+Since in Sophia ML we don’t have `.count` or `.length` to get the list length, we need to make ourselves a helper function which will make a recursion and will iterate over the list while incrementing a counter.
+
+The `size` function is defined to accept a list of `'a` which is the convention for a generic type, and a return type `int` . In the function’s body we are calling the `size'` function, while passing the list and an initial value for the counter.
+``````
+private function size'(l : list('a), x : int) : int =
+      switch(l)
+         [] => x
+         _ :: l' => size'(l', x + 1)
+``````
+And here the magic happens, we use the `switch` statement with 2 cases  `[] => x` – which returns the value of the counter and breaks the recursion if the list is empty. And `_ :: l' => size'(l', x+1)` – means we are using a pattern matching and we are separating the first element from the list and the remainder and then recursively passing the list’s remainder to the same function, while incrementing the counter.
+
+The whole smart contract code looks like this in the end:
+
+```````
+contract Vote =
+
+   record candidates = {
+      voters: list(address) }
+
+   record state = {
+      votes: map(address, candidates) }
+
+   public stateful function init() : state = {
+      votes = { } }
+
+   public stateful function vote(candidate: address) : bool =
+      is_candidate(candidate)
+      let new_votes_state = Call.caller :: state.votes[candidate].voters
+      put(state{
+         votes[candidate].voters = new_votes_state })
+      true
+
+   public function count_votes(candidate : address) : int =
+      size(state.votes[candidate].voters)  
+
+   private function size(l : list('a)) : int = size'(l, 0)
+   
+   private function size'(l : list('a), x : int) : int =
+      switch(l)
+         [] => x
+         _ :: l' => size'(l', x + 1)
+
+   public stateful function add_candidate(candidate: address) : bool =
+      is_candidate(candidate)
+      true
+
+   private stateful function is_candidate(candidate: address) =
+      let candidate_found = lookupByAddress(candidate, state.votes, { voters = [] })
+      if (size(candidate_found.voters) == 0)
+         put(state{
+            votes[candidate] = { voters = [] } })
+
+   private function lookupByAddress(k : address, m, v) =
+      switch(Map.lookup(k,m))
+         None => v
+         Some(x) => x
+`````````
+
+## Conclusion
+It is fairly simple to create a basic aepp on æternity blockchain using Sophia ML. In our case with the ae-vote we stumbled upon some tricky parts like the recursive iteration, we had to make above, but when you familiarize yourself with the language it is easier.
+         
```

Comment:
can you add a reference to ask in the forum for help in case there are problems, for example see the other tutorials